packages/editor/src/lib/editor/Editor.ts

```diff
@@ -5247,13 +5242,24 @@ export class Editor extends EventEmitter<TLEventMap> {
 		return this.updateBindings([partial])
 	}
 
-	deleteBindings(bindings: (TLBinding | TLBindingId)[]) {
+	deleteBindings(bindings: (TLBinding | TLBindingId)[], { isolateShapes = false } = {}) {
 		const ids = bindings.map((binding) => (typeof binding === 'string' ? binding : binding.id))
-		this.store.remove(ids)
+		if (isolateShapes) {
+			for (const id of ids) {
+				const binding = this.getBinding(id)
+				if (!binding) continue
+				const util = this.getBindingUtil(binding)
+				util.onBeforeIsolateFromShape?.({ binding, shape: this.getShape(binding.fromId)! })
+				util.onBeforeIsolateToShape?.({ binding, shape: this.getShape(binding.toId)! })
+				this.store.remove([id])
+			}
```

Comment:
let's wrap this is `this.store.atomic` so that the `after` callbacks happen in the same way they do even without setting `isolateShapes`

Reply:
Hm that cause bugs if the binding's onAfterDelete does some important cleanup that the `onBeforeIsolate` callbacks need to function correctly. Is there any danger of something going wrong how it is?

Reply:
If that's a source of bugs, then surely it would be a source of bugs for normal deletion too, right? When doing normal 1-sided deletion, these callbacks still get called, but the after-phase callbacks won't happen until everything has been deleted. I think you're more likely to end up with bugs as a result of two different calling patterns than a uniform one, considering that's how all the other after callbacks work.

Reply:
huh yeah good point. alright I'll make this atomic and we can revisit the whole business if anybody complains.